design.md

```diff
@@ -0,0 +1,250 @@
+
+## Security - Authentication
+As we've shown above, we'll use the TLS mutual auth to authenticate the client and use the CommonName as the User Identity.
+
+## Security - Authorization
+Our middleware will handle Authorization.
+1. When creating a job, it will store the ID of the job into the map with the User ID as the key.
+2. When requesting access or mutation of a job, we will have middleware that checks the job ID in the param and ensures the ID matches one that the User has access to.
+
+## Router & Endpoints:
+
+### `/api/job [post]`
+This endpoint creates a new Job.
+
+### Payload requirements: 
+```graphql
+{ "cmd": !String, "args": ![String]}
+```
+
+### Response Payload - Success (200): 
+```graphql
+{ "id": uuidv4}
+```
+
+### `/api/job/:jobid [get]`
+This endpoint gets details about the job specified at jobid.
+
+This endpoint will use authorization middleware to check the user has access to the endpoint. If the user doesn't have access, it will send a `403` response and stop routing the request.
+
+### Response Payload - Success (200): 
+```graphql
+{ "cmd": !String, "args": ![String], "active": !bool, "exitCode": int}
+```
+
+### `/api/job/:jobid/output [get]`
+This endpoint gets output regarding the job specified for jobid.
+
+This endpoint will use authorization middleware to check the user has access to the endpoint. If the user doesn't have access, it will send a `403` response and stop routing the request.
+
+### Response Payload - Success (200):
+```graphql
+{
+	active: !bool,
+	output: {
+		stdOut: !String,
+		stdErr: !String,
+	}
+}
+```
+
+### `/api/job/:jobid/stop [post]`
+This sends a kill signal to the command if still running.
+
+This endpoint will use authorization middleware to check the user has access to the endpoint. If the user doesn't have access, it will send a `403` response and stop routing the request.
+
+### Response Payload - Success: 
+```json
+{ "status_code": 202 }
+```
+
+## Tradeoffs
+ - Most of the critical tradeoffs here are around how we handle the use of a self-signed Root Certificate created from the openssl utility.
+
+# Client
+The client will have 4 key commands, and will be named `[Jobs]`.
+1. Start
+2. List
+3. Stop
+5. Output
+
+```bash
+# [] specifies params, ! specifies non-empty constraint
+# Command 1
+$ Jobs start [!cmd] [arg...]
+# CMD Output Happypath
+> ID: F08B2593-FE31-4A37-BF64-E1B584AEB255
+
+$ Jobs start [!cmd] [arg...]
+> ID: DFB531CF-C112-4A11-8016-3AF8D401D347
+
+# Command 2
+$ Jobs list	[id]
+>
+ID			 							Command			Status
+DFB531CF-C112-4A11-8016-3AF8D401D347 	curl			active
+F08B2593-FE31-4A37-BF64-E1B584AEB255	wget			active
+
+$ Jobs list F08B2593-FE31-4A37-BF64-E1B584AEB255
+>
+ID			 							Command			Status
+F08B2593-FE31-4A37-BF64-E1B584AEB255	wget			active
+
+# Command 3
+$ Jobs stop [!id]
+> Stopped # Happypath
+
+```
+
+We will be storing a local "cache" of the values from the server when we know the job has completed.
```

Comment:
Wonder why this is needed. How you make sure that client cache is in sync ?

Reply:
By cache, I really just meant a local storage of jobs output that have already exited and output retrieved. In those cases, there really isn't a possibility for the value on the server to mutate. All state changes to it have happened.